### ISSUE TITLE
pgd_attack should have a better name / documentation

### ISSUE THREAD

**Issue by goodfeli** [tensorflow/cleverhans#780]
If you look at `pgd_attack` in isolation it looks like it's what you should run if you want to use a PGD attack. The doc should at the very least steer people to the ProjectedGradientDescent class. I'm not 100% sure what's going on with this function, but the doc should specify that `pgd_attack` is mostly intended for implementing SPSA, if that's correct, and if that's not correct the doc should explain what actually is going on. Also, I suspect the name `pgd_attack` ought to be changed. For one thing, it should be different to avoid steering users away from ProjectedGradientDescent, but also I think a large part of the point of this function is that it doesn't need to literally use the gradient, so the "g" seems wrong. Maybe it should just be called `projected_optimizer` to reflect that it's relatively generic and reflect that it's not an `Attack`?

**Comment by juesato** [comment]
I originally implemented `pgd_attack` to allow easily implementing a large family of projected gradient attacks (including PGD, with modifications such as e.g. the type of projection operation, the Optimizer class, steepest descent in L2 vs Linf space, whether to use analytic or finite difference gradients, etc.).

It definitely looks strange in context, because it's only used for SPSA in cleverhans. I think the easiest thing to do would be to just add to the docstring that this is a helper function only used in SPSA, and that `ProjectedGradientDescent` implements the PGD attack. WDYT?

**Comment by goodfeli** [comment]
There are 2 things I think we should improve:

1) Having the word `attack` in the name  is strange when there is an `Attack` class and this is not a direct implementation of one of those classes. In the context of this module, it's especially confusing, because many of the other functions in this model really are just the implementation of some `Attack` class. So I would advocate for finding a new name that doesn't have `attack` in it.

2) This is less of a priority for me, and definitely more pedantic, but having `g` for gradient in the name seems to sell short what this does. A lot of its purpose is to optimize functions where the gradient is not available or isn't useful for optimization. So I would advocate for a name that doesn't have the word `gradient` or a `g` that stands for `gradient` in it.

**Comment by juesato** [comment]
Re 1: That makes sense to me. How do you feel about `pgd_attack_helper`? It is an "attack" in the sense that it takes a clean image Tensor (+ some other stuff) and returns an adversarial image Tensor. But I agree it's confusing, esp given most other methods with `attack` in the name are implementations of an `Attack` class.
Other name options are `pgd_helper` or `make_pgd_adv_image`.

2 makes sense to me, but all the other names that stand out to me are really verbose, and more confusing, like `projected_iterative_optimization`, so I'm going to leave it for now.

**Comment by goodfeli** [comment]
`pgd_attack_helper` and`pgd_helper`, still makes me think it's the `attacks_tf` implementation of `ProjectedGradientDescent`, the same way that `fgm` was the `attacks_tf` implementation of `FastGradientMethod`. `make_pgd_adv_image` makes me think you're going to call `ProjectedGradientDescent` and then return the result as a PIL Image or save it as a png file or something. `pgd_attack_recipe` makes me think it's calling `ProjectedGradientDescent` with some hyperparameters that are known to be good.

I'm not trying to be contrarian or push for a really specific name; I'm fine with anything that doesn't make it sound like this is related to the `ProjectedGradientDescent` class.

Here is a brainstormed list of 10 names I think would be better than the current name:
1. `projected_optimizer`
2. `generic_optimizer`
3. `optimization_based_attack`
4. `approximate_gradient_attack`
5. `spsa_helper`
6. `${foo}_attack` where ${foo} is the lower-case underscore version of whatever we name `UnrolledOptimizer`. e.g. `unrolled_optimizer_attack`
7. `run_attack_optimizer`
8. `generalized_pgd`
9. `approximate_pgd`
10. `run_generalized_pgd`

**Comment by juesato** [comment]
Haha okay, I'll rename it to `projected_optimization` or `optimization_based_attack`.

I'll wait until after merging https://github.com/tensorflow/cleverhans/pull/794 to avoid conflicts.
